src/managed/Microsoft.NET.HostModel/AppHost/HostWriter.cs

```diff
@@ -43,45 +44,70 @@ public static class HostWriter
+                        if (windowsGraphicalUserInterface)
                         {
-                            throw new AppHostNotPEFileException();
+                            if (!appHostIsPEImage)
+                            {
+                                throw new AppHostNotPEFileException();
+                            }
+
+                            BinaryUtils.SetWindowsGraphicalUserInterfaceBit(accessor);
                         }
+                    }
+                }
 
-                        BinaryUtils.SetWindowsGraphicalUserInterfaceBit(accessor);
+                if (assemblyToCopyResorcesFrom != null && appHostIsPEImage)
+                {
+                    if (ResourceUpdater.IsSupportedOS())
+                    {
+                        // Copy resources from managed dll to the apphost
+                        new ResourceUpdater(appHostDestinationFilePath)
+                            .AddResourcesFromPEImage(assemblyToCopyResorcesFrom)
+                            .Update();
+                    }
+                    else
+                    {
+                        throw new AppHostCustomizationUnsupportedOSException();
                     }
                 }
-            }
 
-            if (assemblyToCopyResorcesFrom != null && appHostIsPEImage)
+                // Memory-mapped write does not updating last write time
+                File.SetLastWriteTimeUtc(appHostDestinationFilePath, DateTime.UtcNow);
+            }
+            catch (Exception ex)
             {
-                if (ResourceUpdater.IsSupportedOS())
+                FailedToDeleteApphostException failedToDeleteApphostException = null;
+                // Delete the destination file so we don't leave an unmodified apphost
+                try
                 {
-                    // Copy resources from managed dll to the apphost
-                    new ResourceUpdater(appHostDestinationFilePath)
-                        .AddResourcesFromPEImage(assemblyToCopyResorcesFrom)
-                        .Update();
+                    File.Delete(appHostDestinationFilePath);
                 }
-                else 
+                catch (Exception failedToDeleteEx) when (failedToDeleteEx is IOException || failedToDeleteEx is UnauthorizedAccessException)
                 {
-                    throw new AppHostCustomizationUnsupportedOSException();
+                    failedToDeleteApphostException = new FailedToDeleteApphostException(failedToDeleteEx.Message);
                 }
-            }
 
-            // Memory-mapped write does not updating last write time
-            File.SetLastWriteTimeUtc(appHostDestinationFilePath, DateTime.UtcNow);
+                if (failedToDeleteApphostException != null)
+                {
+                    throw new AggregateException(ex, failedToDeleteApphostException);
+                }
+                else
+                {
+                    throw ex;
```

Comment:
`throw; `to preserve stack trace.

Reply:
I don't know i can just throw, even i give exception a name.

although do you think the aggregate exception way is ok? I cannot find a better to surface 2 exception at the same time

Reply:
Also, I'd remove the `else` since the if throws. I like to see the `throw` without conditions so that it's clear that we are handling some extra cases and then letting the original exception propagate.

Reply:
I think AggregateException is OK.

Reply:
changed as suggested